File: blog/convert.py
#!/usr/bin/env python3
"""Convert the Interpretable ML book (Quarto .qmd chapters) into a minimal OKF bundle.
Part -> chapters structure taken verbatim from the book's _quarto.yml."""
import os, re, shutil

SRC, OUT, TS = "src/manuscript", "iml-okf", "2025-04-13T00:00:00Z"
PARTS = {
    "Foundations": ["intro", "interpretability", "goals", "overview", "data"],
    "Interpretable Models": ["limo", "logistic", "extend-lm", "tree", "rules", "rulefit"],
    "Local Model-Agnostic Methods": ["ceteris-paribus", "ice", "lime", "counterfactual", "anchors", "shapley", "shap"],
    "Global Model-Agnostic Methods": ["pdp", "ale", "interaction", "decomposition", "feature-importance", "lofo", "global", "proto"],
    "Neural Network Interpretation": ["cnn-features", "pixel-attribution", "detecting-concepts", "adversarial", "influential"],
    "Beyond the Methods": ["evaluation", "storytime", "future", "translations"],
    "Back Matter": ["cite", "acknowledgements"],
    "Appendix": ["what-is-machine-learning", "math-terms", "r-packages", "references"],
}

shutil.rmtree(OUT, ignore_errors=True); os.makedirs(OUT)
pslug = lambda p: "part-" + re.sub(r"[^a-z0-9]+", "-", p.lower())
yqt = lambda s: '"' + s.replace('"', "'") + '"'

def clean(lines):
    out = []
    for ln in lines:
        if re.match(r"^\s*(\{\{<|:::)", ln):           # drop Quarto directives
            continue
        ln = re.sub(r"\s*\{#[^}]+\}\s*$", "", ln)      # strip {#label} from headings
        ln = re.sub(r"!\[[^]]*\]\([^)]*\)", "", ln)    # remove image embeds
        ln = re.sub(r"\[([^]]+)\]\([^)]*\)", r"\1", ln) # de-link to prose
        out.append(ln)
    return out

def first_sentence(body):
    for ln in body:
        if ln.strip() and not ln.startswith("#"):
            return re.sub(r"([.!?]).*$", r"\1", ln.strip())[:200]
    return ""

def fm(type_, title, desc):
    return ["---", f"type: {type_}", f"title: {yqt(title)}", f"description: {yqt(desc)}",
            f"timestamp: {TS}", "tags: [interpretable-ml]", "---", ""]

def write(name, lines):
    with open(os.path.join(OUT, name), "w", encoding="utf-8", newline="\n") as fh:
        fh.write("\n".join(lines) + "\n")

for part, stubs in PARTS.items():
    for stub in stubs:
        f = os.path.join(SRC, stub + ".qmd")
        if not os.path.exists(f):
            continue
        raw = open(f, encoding="utf-8").read().splitlines()
        h1 = next((l for l in raw if re.match(r"^#\s", l)), None)
        title = stub if h1 is None else re.sub(r"\{#[^}]+\}", "", re.sub(r"^#\s+", "", h1)).strip()
        body = clean(raw)
        write(stub + ".md", fm(part, title, first_sentence(body)) + body)

for part, stubs in PARTS.items():
    links = [f"- [{s}]({s}.md)" for s in stubs]
    write(pslug(part) + ".md", fm("Part", part, f"Chapters in the {part} part.") + [f"# {part}", ""] + links)

ix = fm("Index", "Interpretable Machine Learning",
        "A Guide for Making Black Box Models Explainable - Christoph Molnar (CC BY-NC-SA).")
ix += ["# Interpretable Machine Learning", ""] + [f"- [{p}]({pslug(p)}.md)" for p in PARTS]
write("index.md", ix)
write("log.md", ["---", "type: Log", 'title: "Change log"', f"timestamp: {TS}", "---", "",
                 "# Change log", "", "- Converted from the Quarto sources of *Interpretable Machine Learning*."])
print("wrote", len([f for f in os.listdir(OUT) if f.endswith(".md")]), "OKF files")

File: blog/convert.R
# Convert the Interpretable ML book (Quarto .qmd chapters) into a minimal OKF bundle.
# Part -> chapters structure taken verbatim from the book's _quarto.yml.
src <- "src/manuscript"; out <- "iml-okf"; ts <- "2025-04-13T00:00:00Z"
unlink(out, recursive = TRUE); dir.create(out)

parts <- list(
  "Foundations" = c("intro","interpretability","goals","overview","data"),
  "Interpretable Models" = c("limo","logistic","extend-lm","tree","rules","rulefit"),
  "Local Model-Agnostic Methods" = c("ceteris-paribus","ice","lime","counterfactual","anchors","shapley","shap"),
  "Global Model-Agnostic Methods" = c("pdp","ale","interaction","decomposition","feature-importance","lofo","global","proto"),
  "Neural Network Interpretation" = c("cnn-features","pixel-attribution","detecting-concepts","adversarial","influential"),
  "Beyond the Methods" = c("evaluation","storytime","future","translations"),
  "Back Matter" = c("cite","acknowledgements"),
  "Appendix" = c("what-is-machine-learning","math-terms","r-packages","references")
)

pslug <- function(p) paste0("part-", gsub("[^a-z0-9]+", "-", tolower(p)))
clean <- function(x) {
  x <- x[!grepl("^\\s*(\\{\\{<|:::)", x)]                       # drop Quarto directives
  x <- sub("\\s*\\{#[^}]+\\}\\s*$", "", x)                      # strip {#label} from headings
  x <- gsub("!\\[[^]]*\\]\\([^)]*\\)", "", x)                   # remove image embeds
  gsub("\\[([^]]+)\\]\\([^)]*\\)", "\\1", x)                    # de-link to prose (concept links come from structure)
}
first_sentence <- function(body) {
  p <- body[nzchar(trimws(body)) & !grepl("^#", body)]
  if (!length(p)) return("")
  substr(sub("([.!?]).*$", "\\1", trimws(p[1])), 1, 200)
}
yqt <- function(s) paste0("\"", gsub("\"", "'", s), "\"")
fm <- function(type, title, desc) c("---", paste0("type: ", type),
  paste0("title: ", yqt(title)), paste0("description: ", yqt(desc)),
  paste0("timestamp: ", ts), "tags: [interpretable-ml]", "---", "")

for (part in names(parts)) for (stub in parts[[part]]) {
  f <- file.path(src, paste0(stub, ".qmd")); if (!file.exists(f)) next
  raw <- readLines(f, warn = FALSE, encoding = "UTF-8")
  h1 <- raw[grepl("^#\\s", raw)][1]
  title <- if (is.na(h1)) stub else trimws(sub("\\{#[^}]+\\}", "", sub("^#\\s+", "", h1)))
  body <- clean(raw)
  writeLines(c(fm(part, title, first_sentence(body)), body),
             file.path(out, paste0(stub, ".md")), useBytes = TRUE)
}

# One page per book part, linking its chapters -> a clustered (not flat-star) graph.
for (part in names(parts)) {
  links <- sprintf("- [%s](%s.md)", parts[[part]], parts[[part]])
  writeLines(c(fm("Part", part, paste("Chapters in the", part, "part.")),
               paste0("# ", part), "", links),
             file.path(out, paste0(pslug(part), ".md")), useBytes = TRUE)
}

# index.md links to the parts (the parts link to chapters).
ix <- c(fm("Index", "Interpretable Machine Learning",
           "A Guide for Making Black Box Models Explainable - Christoph Molnar (CC BY-NC-SA)."),
        "# Interpretable Machine Learning", "")
for (part in names(parts)) ix <- c(ix, sprintf("- [%s](%s.md)", part, pslug(part)))
writeLines(ix, file.path(out, "index.md"), useBytes = TRUE)
writeLines(c("---","type: Log","title: \"Change log\"",paste0("timestamp: ", ts),"---","",
             "# Change log","","- Converted from the Quarto sources of *Interpretable Machine Learning*."),
           file.path(out, "log.md"), useBytes = TRUE)
cat("wrote", length(list.files(out, pattern = "\\.md$")), "OKF files\n")
